### PR TITLE
[1 of 2] Introduce router util in integration tests

### PR DESCRIPTION
**Pull request by jfurrow**
This PR introduces a `router` utility that can be used in place of `cy.route`. The `router.route` method will call `cy.route` and store the provided fixture string for future retrieval.

This is useful for writing integration that are agnostic to the API; the tests should work with both fixtures and a live cluster. When asserting that certain values are present, the tests can use this new `router` utility to get either the fixture data or the API response (the latter bit hasn't been built in yet). Then the test can compare the UI's displayed data against the response.

**Checklist**
- [ ] Did you add a JIRA issue in a commit message or as part of the branch name?
- [x] Did you add new unit tests?
- [ ] Did you add new integration tests?
- [ ] If this is a regression, did you write a test to catch this in the future?

### CODE REVIEW

**Comment by MatApple** [comment]
LGTM without understanding the context/direction 👍  @jfurrow - perhaps we can sync offline!

**Comment by jfurrow** [comment]
Sorry @MatApple, I updated the description to provide more details. I'm happy to discuss offline too!

File: tests/_support/utils/router.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    return this;
+  },
+
+  getAPIResponse(routeString, callback) {
+    let selectedFilterString = null;
+
+    _routes.forEach(function (fixtureString, routeRegEx) {
+      if (routeRegEx.test(routeString)) {
+        selectedFilterString = /^fx:(.*)/.exec(fixtureString)[1];
```
Comment (rcorral) [inline tests/_support/utils/router.js]: It's possible to also add fixtures using the `fixture:` prefix, not just `fx:`.
Let's account for that as well!

File: tests/_support/utils/router.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    _routes.set(routeRegEx, fixtureString);
+    cy.route(...arguments);
+
+    return this;
+  },
+
+  getAPIResponse(routeString, callback) {
+    let selectedFilterString = null;
+
+    _routes.forEach(function (fixtureString, routeRegEx) {
```
Comment (rcorral) [inline tests/_support/utils/router.js]: Let's add more comments to this function, it's not super intuitive!

File: tests/_support/utils/router.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+const _routes = new Map();
+
+module.exports = {
```
Comment (rcorral) [inline tests/_support/utils/router.js]: Can you please add some tests for this util?

Comment (jfurrow) [inline tests/_support/utils/router.js]: Done!

File: tests/_support/utils/router.js (hunk 2 above)
Comment (jfurrow) [inline tests/_support/utils/router.js]: Done!

File: tests/_support/utils/router.js (hunk 1 above)
Comment (jfurrow) [inline tests/_support/utils/router.js]: Done!

**Comment by rcorral** [comment]
run integration tests

**Comment by jfurrow** [comment]
@rcorral Sorry, but this actually isn't working as I intended in the subsequent PR, so I'm adding `hold merge` until it's fixed up. Thanks for your patience!

**Comment by rcorral** [comment]
@jfurrow your tests are failing!

**Comment by jfurrow** [comment]
Thanks @rcorral, should be good now. I had to ignore the unit tests in Cypress :)

**Comment by jfurrow** [comment]
Ping @rcorral

File: cypress.json [hunk 4]
```diff
@@ -5,6 +5,7 @@
     "FULL_INTEGRATION_TEST": false
   },
   "fixturesFolder": "tests/_fixtures",
+  "ignoreTestFiles": "**/tests/_support/utils/__tests__/*",
```
Comment (rcorral) [inline cypress.json]: Let's not ignore these tests in cypress. Instead, let's have cypress run them!
Find a way around using jasmine.
https://docs.cypress.io/docs/bundled-tools
Use Sinon: http://sinonjs.org/

File: tests/_support/utils/router.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    _routes.forEach(function (fixtureString, routeRegEx) {
+      if (routeRegEx.test(routeString)) {
+        // The desired fixture string will be the second element in the returned
+        // array.
+        desiredFixtureString = /^(?:fx|fixture):(.*)/.exec(fixtureString)[1];
+      }
+    });
+
+    cy.fixture(desiredFixtureString).then(function (desiredFixture) {
```
Comment (rcorral) [inline tests/_support/utils/router.js]: Should we return early if `desiredFixtureString` is not `null`?

File: tests/_support/utils/router.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      callback(desiredFixture);
+    });
```
Comment (rcorral) [inline tests/_support/utils/router.js]: Can you just do this:
```
cy.fixture(desiredFixtureString).then(callback);
```

File: tests/_support/utils/tests/router-test.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      expect(global.cy.route.calledWith(this.fooRegEx, 'fx:bar')).to.equal(true);
+    });
+
+    it('should return an instance of the router util', function () {
+      expect(this.returnValue).to.equal(router);
+    });
+  });
+
+  describe('#getAPIResponse', function () {
+    it('should return a promise which resolves with the fixture when found', function (done) {
```
Comment (rcorral) [inline tests/_support/utils/tests/router-test.js]: It doesn't return a promise..

File: tests/_support/utils/tests/router-test.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      router.route(/foo/, 'fx:bar');
+      router.getAPIResponse('foo', function (foundFixture) {
+        expect(foundFixture).to.equal('bar');
+        done();
+      });
+    });
+
+    it('should return a promise which resovles with null when fixture is not found', function (done) {
```
Comment (rcorral) [inline tests/_support/utils/tests/router-test.js]: Same thing here ;)

File: tests/_support/utils/router.js (hunk 5 above)
Comment (jfurrow) [inline tests/_support/utils/router.js]: Good call, done!

File: tests/_support/utils/router.js (hunk 6 above)
Comment (jfurrow) [inline tests/_support/utils/router.js]: Yep, done!

File: tests/_support/utils/tests/router-test.js (hunk 7 above)
Comment (jfurrow) [inline tests/_support/utils/tests/router-test.js]: Good catch, fixed!

File: tests/_support/utils/tests/router-test.js (hunk 8 above)
Comment (jfurrow) [inline tests/_support/utils/tests/router-test.js]: Good catch, fixed!

**Pull request merged**